### ISSUE TITLE
Dynamically updating Neumann load boundary conditions for Incremental Loading

### ISSUE THREAD

**Issue by RayKruger** [deepmodeling/jax-fem#52]
I am modeling large-strain linear elasticity for a cantilever beam using JAX-FEM. In this simulation, I need to dynamically update the Neumann boundary conditions (applied loads) over multiple load steps, incrementally increasing the applied force.

In the plasticity example, I see that Dirichlet displacement boundary conditions can easily be updated dynamically in multiple steps using:

`problem.fe.update_Dirichlet_boundary_conditions(dirichlet_bc_info)`

where `dirichlet_bc_info` is structured as:

`dirichlet_bc_info = [location_fns, vecs, value_fns]`

However, the Neumann bc in the problem class currently only accepts `location_fns` (position functions) and does not support value functions, making dynamic load updates a little harder.

I implemented a workaround by using the `problem.get_surface_maps`  with the optional argument `cell_internal_vars_surface` and providing a custom `problem.update_load` method:

```
def get_surface_maps(self):
    def surface_map(u, x, cell_internal_vars_surface):
        return np.array([0.0, cell_internal_vars_surface])
    return [surface_map]
                
def update_load(self, new_load):
    for i in range(len(self.internal_vars_surfaces)):
        num_face_quads = self.physical_surface_quad_points[i].shape[0]
        # Create an array of shape (num_face_quads,) filled with new_load
        self.internal_vars_surfaces[i] = (np.full((num_face_quads,), new_load),)
```
Example of the multi-step Neumann bc I got working:

![Image](https://github.com/user-attachments/assets/fca4a284-75b3-4c2e-ab52-fa7c7c3187e3)

I also had to adjust the internal JAX-FEM `surface_kernel`’s `jnp.vmap` call to match dimensions. This solution is rather hacky and requires digging deep into the internals.

**Is there a better way to do this?**  **Could the problem class be extended to support dynamic Neumann bc function updates similar to the Dirichlet boundary condition**?

**Comment by xwpken** [comment]
Hi, it seems that you want to change the magnitude of the applied forces, right? If so, your implementation is correct. But you don't need to change the internal codes, for we already have defined the same variable `internal_vars_surfaces`, see https://github.com/deepmodeling/jax-fem/blob/8ea5742ed7d8dd6f482b197d154aa442f4209807/jax_fem/problem.py#L134

You only need to set up the `internal_vars_surfaces` with the correct format (corresponding to each `get_surface_map`), and the `surface_kernel` will get the input.

**Comment by GuillaumeBroggi** [comment]
Hi @xwpken, could you elaborate on the correct format of `internal_vars_surfaces`? For instance, if there is one `get_surface_map` expecting one scalar to update the load?

**Comment by xwpken** [comment]
Hi @GuillaumeBroggi, this is an example of how to input the surface parameters:

```python
def set_params(self, params):
    surface_params = params
    # Generally, [[surface1_params1, surface1_params2, ...], [surface2_params1, surface2_params2, ...], ...]
    self.internal_vars_surfaces = [[surface_params]]`
```
The shape of the `surface_params` should be `(num_selected_faces, num_face_quads, dim)`. For a scalar parameter, `dim` will be 1. More information can be found at the API reference https://deepmodeling.github.io/jax-fem/more/api/api_problem.html.

**Comment by GuillaumeBroggi** [comment]
Thanks @xwpken  !

I managed to obtain the proper shape with:
```python
 def get_surface_maps(self):
        def traction(u, x, load):
             return np.array([0.0, 0.0, load[0]])
        return [traction]
```

and 

```python
shape = (len(problem.boundary_inds_list[0]), problem.fes[0].num_face_quads, 1)

problem.internal_vars_surfaces = [
    [
        np.full(fill_value=-1, sahpe=shape),
    ]
]
```

I did not find the information about `(num_selected_faces, num_face_quads, dim)`, maybe it could be added explicitely to https://deepmodeling.github.io/jax-fem/more/api/api_problem.html#jax_fem.problem.Problem.set_params ?

**Comment by xwpken** [comment]
@GuillaumeBroggi Thank you for your advice; we will improve it.
